dlib/dnn/loss.h

```diff
@@ -3301,6 +3302,12 @@ namespace dlib
                 }
             }
 
+#ifdef DLIB_USE_CUDA
+            double loss;
+            cuda_compute(truth, output_tensor, grad, loss);
+            return loss;
+#else
+
```

Comment:
Can you move the stuff in the `else` block into something analogous to the `cuda::compute_loss_mean_squared_per_channel_and_pixel` you added?  Then in the unit tests you can create both the cuda and cpu version, give them the same inputs, and check that both methods give identical results.

Reply:
Sure! Where should I put that? Inside this class or in `dlib/cuda/cpu_dlib.cpp` and then name it `cpu::compute_loss_mean_squared_per_channel_and_pixel`?

Reply:
OK, I think I've done something sensible in the meantime (I think we have 13 hours of time difference).

I don't get identical results for cuda and cpu implementations. However, the difference between them is around 1e-7 or 1r-8. For this reason, I added a small tolerance margin in the tests of 1e-6.

Reply:
Sounds good